cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcProvisioner.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.runtime.spi.provisioner.dataproc;
+
+import co.cask.cdap.runtime.spi.provisioner.Cluster;
+import co.cask.cdap.runtime.spi.provisioner.ClusterStatus;
+import co.cask.cdap.runtime.spi.provisioner.Node;
+import co.cask.cdap.runtime.spi.provisioner.ProgramRun;
+import co.cask.cdap.runtime.spi.provisioner.Provisioner;
+import co.cask.cdap.runtime.spi.provisioner.ProvisionerContext;
+import co.cask.cdap.runtime.spi.provisioner.ProvisionerSpecification;
+import com.google.common.annotations.VisibleForTesting;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Provisions a cluster using GCE DataProc.
+ */
+public class DataProcProvisioner implements Provisioner {
+  private static final ProvisionerSpecification SPEC = new ProvisionerSpecification(
+    "gce-dataproc", "GCE DataProc Provisioner",
+    "Runs programs on the GCE DataProc clusters.",
+    new HashMap<>());
+
+  @Override
+  public ProvisionerSpecification getSpec() {
+    return SPEC;
+  }
+
+  @Override
+  public void validateProperties(Map<String, String> properties) {
+    DataProcConf.fromProperties(properties);
+  }
+
+  @Override
+  public Cluster createCluster(ProvisionerContext context) throws Exception {
+    DataProcConf conf = DataProcConf.fromProperties(context.getProperties());
+    String clusterName = getClusterName(context.getProgramRun());
+
+    try (DataProcClient client = DataProcClient.fromConf(conf)) {
+      // if it already exists, it means this is a retry. We can skip actually making the request
+      Optional<com.google.cloud.dataproc.v1.Cluster> existing = client.getCluster(clusterName);
+
+      // TODO: figure out how to get this if existing is not null
```

Comment:
What is the TODO for?

Reply:
this situation is if we're retrying this call and the cluster already exists, then we should use the create time for the original create request, not the current time.